doc/api/worker_threads.md

```diff
@@ -1329,6 +1329,228 @@ from the running process and will preload the same preload scripts as the main
+
+`SynchronousWorker` was originally developer by Node.js core contributor
+Anna Henningsen and published as a separate module [`synchronous-worker`][] on
+npm under the MIT license. It was integrated into Node.js core with Anna's
+permission. The majority of the code, documentation, and tests were adopted
+almost verbatim from the original module.
+
+#### Why would I use a SynchronousWorker?
+
+The most common use case is probably running asynchronous code synchronously,
+in situations where doing so cannot be avoided (even though one should try
+really hard to avoid it). Another popular npm package that does this is
+[`deasync`][], but `deasync`
+
+* solves this problem by starting the event loop while it is already running
+  (which is explicitly _not_ supported by libuv and may lead to crashes)
+* doesn’t allow specifying _which_ resources or callbacks should be waited for,
+  and instead allows everything inside the current thread to progress.
+
+#### How can I avoid using SynchronousWorker?
+
+If you do not need to directly interact with the objects inside the inner
+Node.js instance, a lot of the time Worker threads together with
+[`Atomics.wait()`][] will give you what you need.
+
+#### My async functions/Promises/… don’t work
+
+If you run a `SynchronousWorker` with its own microtask queue (i.e. in default
+mode), code like this will not work as expected:
+
+```cjs
+const { SynchronousWorker } = require('node:worker_threads');
+const w = new SynchronousWorker();
+let promise;
+w.runInWorkerScope(() => {
+  promise = (async () => {
+    return w.createRequire(__filename)('node-fetch')('...');
+  })();
+});
+w.runLoopUntilPromiseResolved(promise);
+```
+
+The reason for this is that `async` functions (and Promise `.then()` handlers)
+add their microtasks to the microtask queue for the Context in which the
+async function (or `.then()` callback) was defined, and not the Context in which
+the original `Promise` was created. Put in other words, it is possible for a
+`Promise` chain to be run on different microtask queues.
+
+While this behavior may be counterintuitive, it is what the V8 engine does,
+and is not under the control of Node.js.
+
+What this means is that you will need to make sure that the functions are
+compiled in the Context in which they are supposed to be run; the two main
+ways to achieve that are to:
+
+* Put them in a separate file that is loaded through `w.createRequire()`
+* Use `w.createRequire(__filename)('vm').runInThisContext()` to manually compile
+  the code for the function in the Context of the target Node.js instance.
+
+For example:
+
+```cjs
+const { SynchronousWorker } = require('node:worker_threads');
+const w = new SynchronousWorker();
+const req = w.createRequire(__filename);
+let promise;
+w.runInWorkerScope(() => {
+  promise = req('vm').runInThisContext(`(async(req) => {
+    return await req('node-fetch')('...');
+  })`)(req);
+});
+w.runLoopUntilPromiseResolved(promise);
+```
```

Comment:
Adding ESM version; does `w.createRequire` accept `import.meta.url`? Could this be an `import` instead, because I assume the generated `require` only accepts CJS code as input?

```suggestion
```mjs
import { SynchronousWorker } = from 'node:worker_threads';
const w = new SynchronousWorker();
const req = w.createRequire(import.meta.url);
let promise;
w.runInWorkerScope(() => {
  promise = req('node:vm').runInThisContext(`(async(req) => {
    return await req('node-fetch')('...');
  })`)(req);
});
w.runLoopUntilPromiseResolved(promise);
```
```cjs
const { SynchronousWorker } = require('node:worker_threads');
const w = new SynchronousWorker();
const req = w.createRequire(__filename);
let promise;
w.runInWorkerScope(() => {
  promise = req('node:vm').runInThisContext(`(async(req) => {
    return await req('node-fetch')('...');
  })`)(req);
});
w.runLoopUntilPromiseResolved(promise);
```
```